### DISCUSSION TITLE
prefetch_count seems to consume always only 1 message

### DISCUSSION THREAD

**Discussion by CodeMiner84** [php-amqplib/php-amqplib#920] [Q&A] [answered]
**Description**  
According to this tutorial:
https://www.rabbitmq.com/tutorials/tutorial-two-php.html

prefetch_count = 1 means not to give more than one message to a worker at a time.

so having for example prefetch_count = 10 in should consume 10 messages at a time, correct ?

**How to reproduce**  
`new_task.php`

```php
<?php

require_once __DIR__ . '/vendor/autoload.php';
use PhpAmqpLib\Connection\AMQPStreamConnection;
use PhpAmqpLib\Message\AMQPMessage;

$connection = new AMQPStreamConnection('localhost', 5672, 'guest', 'guest');
$channel = $connection->channel();

$channel->queue_declare('task_queue', false, true, false, false);

$data = implode(' ', array_slice($argv, 1));
if (empty($data)) {
    $data = "Hello World!";
}
$msg = new AMQPMessage(
    $data,
    array('delivery_mode' => AMQPMessage::DELIVERY_MODE_PERSISTENT)
);

$channel->basic_publish($msg, '', 'task_queue');

echo ' [x] Sent ', $data, "\n";

$channel->close();
$connection->close();
```

`worker.php`
```php
<?php

require_once __DIR__ . '/vendor/autoload.php';
use PhpAmqpLib\Connection\AMQPStreamConnection;

$connection = new AMQPStreamConnection('localhost', 5672, 'guest', 'guest');
$channel = $connection->channel();

$channel->queue_declare('task_queue', false, true, false, false);

echo " [*] Waiting for messages. To exit press CTRL+C\n";

$callback = function ($msg) {
    echo ' [x] Received ', $msg->body, "\n";
    echo " [x] Done\n";
    $msg->ack();

    sleep(2);
};

$channel->basic_qos(null, 1, null);
$channel->basic_consume('task_queue', '', false, false, false, false, $callback);

while ($channel->is_open()) {
    $channel->wait();
}

$channel->close();
$connection->close();
```
I've only add `sleep(2);` function to callback to simulate delay in a process. The rest is from the tutorial

So first start worker by `php worker.php` and in other terminal hit 10 times `php new_task.php` to produce 10 messages.

If one message is consumed in 2 seconds - whole process for prefetch_count = 1 should took about 20 seconds.

But with prefetch_count = 10 it should be able to consume 10 messages at same time so ~ 2sec - am I correct ? One worker ? Not 10 running workers in backgroud. Or maybe prefetch_count is completely for other things and this cannot be accomplished ?

**Comment by ramunasd** [comment, marked as answer]
> so having for example prefetch_count = 10 in should consume 10 messages at a time, correct ?

This assumption is incorrect. prefetch_count determines how many messages at most must be assigned and sent for each individual consumer. This is kind of performance optimization for broker and consumer, because it allows to reduce network roundtrips and spend less time on message distribution.

On library side, messages always comes one after one. Once You finish one message, library gets another one from network buffer(if there are) or waits for network activity. So whole processing is always sequential.

In theory, You could consume messages in parallel using just one system process. But such approach would need support for real threads(unlikely you have it) and a lot of rework in library itself. It's much more easier to run several system processes rather than several threads.

Cheers
